mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergPartitionTextParser.java

```
@@ -0,0 +1,112 @@
+import java.util.regex.Pattern;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.exceptions.ValidationException;
+
+/**
+ * Parser for inputs received from config {@link org.apache.iceberg.mr.InputFormatConfig#PARTITIONING}
+ */
+class HiveIcebergPartitionTextParser {
+  private static final int PARTITION_TEXT_MAX_LENGTH = 1000;
+
+  private static final Pattern BUCKET_PATTERN = Pattern.compile("bucket\\((.*?),(.*?)\\)", Pattern.CASE_INSENSITIVE);
+  private static final Pattern TRUNCATE_PATTERN = Pattern.compile(
+      "truncate\\((.*?),(.*?)\\)", Pattern.CASE_INSENSITIVE);
+  private static final Pattern YEAR_PATTERN = Pattern.compile("year\\((.*?)\\)", Pattern.CASE_INSENSITIVE);
+  private static final Pattern MONTH_PATTERN = Pattern.compile("month\\((.*?)\\)", Pattern.CASE_INSENSITIVE);
+  private static final Pattern DAY_PATTERN = Pattern.compile("day\\((.*?)\\)", Pattern.CASE_INSENSITIVE);
+  private static final Pattern HOUR_PATTERN = Pattern.compile("hour\\((.*?)\\)", Pattern.CASE_INSENSITIVE);
+  private static final Pattern ALWAYS_NULL_PATTERN = Pattern.compile(
+      "alwaysNull\\((.*?)\\)", Pattern.CASE_INSENSITIVE);
+  private static final String INTEGER_PATTERN = "\\d+";
+  private static final String PIPE_DELIMITER = "\\|";
+
+  private HiveIcebergPartitionTextParser() {
+  }
+
+  private static boolean build2ArgTransform(Pattern transformPattern, Schema schema,
+                                         BiConsumer<String, Integer> transformBuilder, String trimmedPart) {
+    Matcher matcher = transformPattern.matcher(trimmedPart);
+    if (matcher.find()) {
+      ValidationException.check(matcher.groupCount() == 2 && matcher.group(2).trim().matches(INTEGER_PATTERN),
+          "Cannot parse 2-arg partition transform from text part: %s", trimmedPart);
+
+      String columnName = matcher.group(1).trim();
+      ValidationException.check(schema.findField(columnName) != null,
```

Nit: I think we can delegate such validation exception to partition spec builder so that we don't need to test ourselves? That has to catch more complicated cases like unexpected column types anyway

---

Yeah agree, these can always be caught by the partition spec builder, but I was thinking it might be better to fail fast. Let me think about it more for what is the best line to cut here.